java/jar/jar_test.go

```diff
@@ -259,6 +259,98 @@ func TestJARBadManifest(t *testing.T) {
 	}
 }
 
+// TestMalformed creates a malformed zip, then makes sure the package handles it
+// gracefully.
+func TestMalformed(t *testing.T) {
+	const (
+		jarName  = `malformed_zip.jar`
+		manifest = `testdata/malformed_zip.MF`
+	)
+	t.Parallel()
+	ctx := zlog.Test(context.Background(), t)
+	dir := integration.PackageCacheDir(t)
+	fn := filepath.Join(dir, jarName)
+Open:
+	f, err := os.Open(fn)
+	switch {
+	case errors.Is(err, nil):
+	case errors.Is(err, os.ErrNotExist):
+		// Create the jar-like.
+		mk, err := os.Create(fn)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer func() {
+			if err := mk.Close(); err != nil {
+				t.Logf("non-failing error: %v", err)
+			}
+			if t.Failed() {
+				if err := os.Remove(fn); err != nil {
+					t.Error(err)
+				}
+			}
+		}()
+		w := zip.NewWriter(mk)
+		if _, err := w.Create(`META-INF/`); err != nil {
+			t.Fatal(err)
+		}
+		fw, err := w.Create(`META-INF/MANIFEST.MF`)
+		if err != nil {
+			t.Fatal(err)
+		}
+		mf, err := os.ReadFile(manifest)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if _, err := io.Copy(fw, bytes.NewReader(mf)); err != nil {
+			t.Fatal(err)
+		}
+		if err := w.Close(); err != nil {
+			t.Fatal(err)
+		}
+
+		// Then, corrupt it.
+		// Seek to the central directory footer:
+		pos, err := mk.Seek(-0x16+0x10 /* sizeof(footer) + offset(dir_offset)*/, io.SeekEnd)
+		if err != nil {
+			t.Fatal(err)
+		}
+		b := make([]byte, 4)
+		if _, err := io.ReadFull(mk, b); err != nil {
+			t.Fatal(err)
+		}
+		// Offset everything so the reader slowly descends into madness.
+		b[0] -= 7
+		if _, err := mk.WriteAt(b, pos); err != nil {
+			t.Fatal(err)
+		}
+
+		if err := mk.Sync(); err != nil {
+			t.Error(err)
+		}
+		goto Open
+	default:
+		t.Fatal(err)
+	}
+	defer f.Close()
+	fi, err := f.Stat()
+	if err != nil {
+		t.Fatal(err)
+	}
+	z, err := zip.NewReader(f, fi.Size())
+	if err != nil {
+		t.Fatal(err)
+	}
+	infos, err := Parse(ctx, jarName, z)
```

Comment:
Should we check the error type here as it is important for how the `PackageScanner` proceeds?

Reply:
as in, make sure it's `NotAJar`? sure